File: src/lib/acceso-turno.ts
/** RNF-A02 — Acceso por horario del turno asignado al empleado. */
const ROLES_SIN_RESTRICCION = ["Administrador", "Recursos Humanos"];

type TurnoHorario = { horaInicio: string; horaFin: string; nombre?: string };

function parseMinutos(hhmm: string): number {
  const [h, m] = hhmm.split(":").map(Number);
  return h * 60 + (m || 0);
}

function minutosActuales(): number {
  const now = new Date();
  return now.getHours() * 60 + now.getMinutes();
}

export function enVentanaTurno(turno: TurnoHorario): boolean {
  const start = parseMinutos(turno.horaInicio);
  const end = parseMinutos(turno.horaFin);
  const now = minutosActuales();

  if (start <= end) {
    return now >= start && now < end;
  }
  return now >= start || now < end;
}

export function mensajeFueraDeTurno(): string {
  return "El sistema no se encuentra disponible para su turno en este momento.";
}

export function rolExentoHorarioLaboral(rol: string): boolean {
  return ROLES_SIN_RESTRICCION.includes(rol);
}

File: backend/core/services/acceso_turno.py
ROLES_SIN_RESTRICCION = ["Administrador", "Recursos Humanos"]


def parse_minutos(hhmm: str) -> int:
    h, m = hhmm.split(":")
    return int(h) * 60 + int(m or 0)


def minutos_actuales() -> int:
    from datetime import datetime

    now = datetime.now()
    return now.hour * 60 + now.minute


def en_ventana_turno(turno) -> bool:
    start = parse_minutos(turno.horaInicio)
    end = parse_minutos(turno.horaFin)
    now = minutos_actuales()
    if start <= end:
        return start <= now < end
    return now >= start or now < end


def mensaje_fuera_de_turno() -> str:
    return "El sistema no se encuentra disponible para su turno en este momento."


def rol_exento_horario_laboral(rol: str) -> bool:
    return rol in ROLES_SIN_RESTRICCION
